pht_federated/aggregator/services/discovery/column_harmonization.py

```diff
@@ -0,0 +1,217 @@
+    )
+
+    return difference_report
+
+
+def create_difference_report(
+    type_differences: List[List[Tuple[str, str]]],
+    column_value_differences: List[Tuple[str, str]],
+    column_value_differences2: List[Tuple[str, str]],
+    column_name_differences: List[List[Tuple[str, str]]],
+    dataset_name: str,
+) -> dict:
+    """
+    Transforms multiple types of mismatch errors between datasets into a summarized difference report
+    :param type_differences: List[List[Tuple[col_name_df, type], Tuple[col_name_aggregator, type]]] differences in type
+    :param column_value_differences: List[Tuple[col_name, type]] differences (Dataframe - Aggregator)
+    :param column_value_differences2: List[Tuple[col_name, type]] differences (Aggregator - Dataframe)
+    :param column_name_differences: List[List[Tuple[col_name_df, type], Tuple[col_name_aggregator, type]]] differences
+                                    in name only
+    :param dataset_name: String that defines the name of local dataset
+    :return: Dictionary which lists the differences between the two datasets
+    """
+
+    mismatch_errors_list = []
+
+    difference_report = {
+        "dataset": dataset_name,
+        "datatype": "tabular",
+        "status": "",
+        "errors": [],
+    }
+
+    for diff in type_differences:
+        case = {
+            "column_name": diff[0][0],
+            "error": {
+                "type": "type",  # missing, type, semantic, extra
+                "dataframe_type": diff[0][1],
+                "aggregator_type": diff[1][1],
+            },
+            "hint": f"Change type to {diff[1][1]}",
+        }
+        mismatch_errors_list.append(case)
+
+    for diff in column_value_differences:
+        case = {
+            "column_name": diff[0],
+            "error": {
+                "type": "added",  # missing, type, semantic, extra
+                "dataframe_type": diff[1],
+            },
+            "hint": f"Column name {diff[0]} only exists in local dataset",
+        }
+        mismatch_errors_list.append(case)
+
+    for diff in column_value_differences2:
+        case = {
+            "column_name": diff[0],
+            "error": {
+                "type": "missing",  # missing, type, semantic, extra
+                "aggregator_type": diff[1],
+            },
+            "hint": f"Column name {diff[0]} only exists in aggregator dataset",
+        }
+        mismatch_errors_list.append(case)
+
+    for diff in column_name_differences:
+        case = {
+            "column_name": diff[0][0],
+            "error": {
+                "type": "added",  # missing, type, semantic, extra
+                "dataset_name": diff[0][0],
+                "aggregator_name": diff[1][0],
+                "aggregator_type": diff[1][1],
+            },
+            "hint": f"Column name {diff[0][0]} only exists in local dataset."
+            f" Did you mean column name: {diff[1][0]}",
+        }
+        mismatch_errors_list.append(case)
```

Comment:
Comment on what is happening here. Seens like a lot could be outsourced into a separate function that creates the case that then gets added to the errors

Reply:
You would also need to differentiate in the extra function again since different keys and strings in each case... is it worth it?